File: Javascript/Strings/checkTwoStringsPermutation.js
// Given two strings, check to see if one is the permutation of the other

function checkTwoStringsPermutation(str1, str2) {
  if (str1.length !== str2.length) {
    return false;
  }
  var map1 = {};
  var map2 = {};
  for (let i = 0; i < str1.length; i++) {
    map1[str1[i]] = map1[str1[i]] + 1 || 1;
  }

  for (let i = 0; i < str2.length; i++) {
    map2[str2[i]] = map2[str2[i]] + 1 || 1;
  }

  if (Object.keys(map1).length !== Object.keys(map2).length) {
    return false;
  }

  for (let key in map1) {
    if (map1[key] !== map2[key]) {
      return false;
    }
  }
  return true;
}

File: Java/Strings/checkTwoStringsPermutation.java
// Given two strings, check to see if one is the permutation of the other

import java.util.HashMap;

public class checkTwoStringsPermutation {
	 boolean checkPermutation(String str1, String str2){
		    if(str1.length() != str2.length()){
		      return false;
		    }
		    HashMap<Character,Integer> hmap1 = new HashMap<Character,Integer>();
		    HashMap<Character,Integer> hmap2 = new HashMap<Character,Integer>();

		    for(int i=0; i<str1.length(); i++){
		      if(hmap1.containsKey(str1.charAt(i))){
		        hmap1.put(str1.charAt(i), hmap1.get(str1.charAt(i))+1);
		      }else{
		        hmap1.put(str1.charAt(i),1);
		      }
		    }
		    for(int i=0; i<str2.length(); i++){
			      if(hmap2.containsKey(str2.charAt(i))){
			        hmap2.put(str2.charAt(i), hmap2.get(str2.charAt(i))+1);
			      }else{
			        hmap2.put(str2.charAt(i),1);
			      }
			    }
		    
		    if(hmap1.keySet().size() != hmap2.keySet().size()){
		    	return false;
		    }
		    
		    for(Character key : hmap1.keySet()){
		    	if(hmap1.get(key) != hmap2.get(key)){
		    		return false;
		    	}
		    }
		    return true;
}
